ext/openssl/openssl.c

```diff
@@ -1399,6 +1399,48 @@ static X509 *php_openssl_x509_from_str(zend_string *cert_str) {
 			return NULL;
 		}
 		cert = PEM_read_bio_X509(in, NULL, NULL, NULL);
+	} else if (ZSTR_LEN(cert_str) > 7 && memcmp(ZSTR_VAL(cert_str), "pkcs11:", sizeof("pkcs11:") - 1) == 0) {
+		char *verbose = NULL;
+		ENGINE *engine;
+	        struct {
+			const char *s_slot_cert_id;
+			X509 *cert;
+		} parms = {
+			.s_slot_cert_id = ZSTR_VAL(cert_str),
+			.cert = NULL,
+		};
+		int force_login = 0;
+
+		engine = ENGINE_by_id("pkcs11");
+		if (engine == NULL) {
+			php_openssl_store_errors();
+			return NULL;
+		}
+		verbose = getenv("OPENSSL_ENGINE_VERBOSE");
+		if (verbose) {
+			if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+				ENGINE_free(engine);
+				php_openssl_store_errors();
+				return NULL;
+			}
+		}
+		if (!ENGINE_init(engine)) {
+			ENGINE_free(engine);
+			php_openssl_store_errors();
+			return NULL;
+		}
```

Comment:
It would be good to wrap this to a function that returns engine. The differences with the below code are just the warnings which could be parametrized.

Reply:
I tried but I felt that there is no benefit: for each setup of the engine, it is used differently.

For the case `php_openssl_pkey_from_zval()`, there are some usages with the macro `TMP_CLEAN`.

For the case `php_openssl_x509_from_str()`, we use:
```
                        php_openssl_store_errors();
                        return NULL;
```

Of course, from a "C" point of view, I could macro-ize or have some tricks to wrap it, but I feel it would add more complexity to maintain than the benefit of wrapping.

Reply:
I think it's pretty much the same except the TMP_CLEAN bit and errors and it would remove the duplication. Maybe something like (not tested - might contain mistakes :) ):
```c
static ENGINE *php_openssl_make_pkcs11_engine(bool show_warnings)
{
                ENGINE *engine = ENGINE_by_id("pkcs11");
		if (engine == NULL) {
                        if (show_warnings) {
                            php_error_docref(NULL, E_WARNING, "Cannot load PKCS11 engine");
                        }
			php_openssl_store_errors();
			return NULL;
		}
		verbose = getenv("OPENSSL_ENGINE_VERBOSE");
		if (verbose) {
			if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
				ENGINE_free(engine);
				php_openssl_store_errors();
				return NULL;
			}
		}
		if (!ENGINE_init(engine)) {
			ENGINE_free(engine);
                        if (show_warnings) {
                            php_error_docref(NULL, E_WARNING, "Cannot init PKCS11 engine");
                        }
			php_openssl_store_errors();
			return NULL;
		}
}
```

And then for `php_openssl_x509_from_str`, it would be just
```c
engine = php_openssl_make_pkcs11_engine(0);
if (engine == NULL) {
   return NULL;
}
```
And for `php_openssl_pkey_from_zval`, it would be
```c
engine = php_openssl_make_pkcs11_engine(1);
if (engine == NULL) {
   TMP_CLEAN;
}
```

Unless I miss some important functionality, it seems better to me. Don't you think? :)

Reply:
Ok you got the point. I'll integrate and update based on your suggestion.

Reply:
done: https://github.com/php/php-src/pull/6860/commits/d7215222fe36c7faa7ca6b612607ad1d51dd6fa8

I  like your design proposal, it avoids the duplication of design errors (ENGINE_free(), ENGINE_finish()).